Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm

```diff
@@ -906,6 +907,78 @@ static bool needsEAGLOnMac()
         m_cv->invalidateKnownTextureContent(texture);
 }
 
+void GraphicsContextGLCocoa::paintRenderingResultsToCanvas(ImageBuffer& imageBuffer)
+{
+    withDrawingBufferAsNativeImage([&](NativeImage& image) {
+        paintToCanvas(image, imageBuffer.backendSize(), imageBuffer.context());
+    });
+}
+
+void GraphicsContextGLCocoa::paintCompositedResultsToCanvas(ImageBuffer& imageBuffer)
+{
+    withDisplayBufferAsNativeImage([&](NativeImage& image) {
+        paintToCanvas(image, imageBuffer.backendSize(), imageBuffer.context());
+    });
+}
+
+void GraphicsContextGLCocoa::withDrawingBufferAsNativeImage(std::function<void(NativeImage&)> func)
+{
+    if (!makeContextCurrent())
+        return;
+
+    if (!m_displayBufferBacking
+        || m_displayBufferBacking->size() != getInternalFramebufferSize())
+        return;
+
+    prepareTexture();
+
+    RetainPtr<CGContextRef> cgContext;
+    RefPtr<NativeImage> drawingImage;
+
+    if (contextAttributes().premultipliedAlpha) {
+        // Use the IOSurface backed image directly
+        EGL_WaitUntilWorkScheduledANGLE(platformDisplay());
```

Comment:
In fact, this should be call to GraphicsContextGLCocoa::waitUntilWorkScheduled()

```
void GraphicsContextGLCocoa::waitUntilWorkScheduled() {
   if (is metal) {
           EGL_WaitUntilWorkScheduledANGLE(platformDisplay());
  } else {
         GL_Flush();
  }
}

```